File: wire-compiler/src/test/java/com/squareup/wire/WireCompilerTest.java
/*
 * Copyright 2013 Square Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire;

import java.io.File;
import java.io.IOException;
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import okio.Okio;
import okio.Source;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import static org.assertj.core.api.Assertions.assertThat;

public class WireCompilerTest {
  @Rule public final TemporaryFolder temp = new TemporaryFolder();

  private StringWireLogger logger;
  private File testDir;

  @Before public void setUp() {
    testDir = temp.getRoot();
  }

  @Test public void testFooBar() throws Exception {
    String[] sources = {
        "foo.proto",
        "bar.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/foobar/protos/bar/Bar.java",
        "com/squareup/foobar/protos/foo/Foo.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testDifferentPackageFooBar() throws Exception {
    String[] sources = {
        "differentpackage/foo.proto",
        "differentpackage/bar.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/differentpackage/protos/bar/Bar.java",
        "com/squareup/differentpackage/protos/foo/Foo.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testPerson() throws Exception {
    String[] sources = {
        "person.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/person/Person.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testPersonAndroid() throws Exception {
    String[] sources = {
        "person.proto"
    };
    invokeCompiler(sources, "--android");

    String[] outputs = {
        "com/squareup/wire/protos/person/Person.java"
    };
    assertOutputs(outputs, ".android");
  }

  @Test public void testPersonCompact() throws Exception {
    String[] sources = {
        "all_types.proto"
    };
    invokeCompiler(sources, "--compact");

    String[] outputs = {
        "com/squareup/wire/protos/alltypes/AllTypes.java"
    };
    assertOutputs(outputs, ".compact");
  }

  @Test public void testSimple() throws Exception {
    String[] sources = {
        "simple_message.proto",
        "external_message.proto",
        "foreign.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/simple/SimpleMessage.java",
        "com/squareup/wire/protos/simple/ExternalMessage.java",
        "com/squareup/wire/protos/foreign/ForeignEnum.java",
        "com/squareup/wire/protos/foreign/ForeignMessage.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testOneOf() throws Exception {
    String[] sources = {
        "one_of.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/oneof/OneOfMessage.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testSingleLevel() throws Exception {
    String[] sources = {
        "single_level.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/single_level/Foo.java",
        "com/squareup/wire/protos/single_level/Foos.java",
    };
    assertOutputs(outputs);
  }

  @Test public void testSameBasename() throws Exception {
    String[] sources = {
        "single_level.proto",
        "samebasename/single_level.proto" };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/single_level/Foo.java",
        "com/squareup/wire/protos/single_level/Foos.java",
        "com/squareup/wire/protos/single_level/Bar.java",
        "com/squareup/wire/protos/single_level/Bars.java",
    };
    assertOutputs(outputs);
  }

  @Test public void testChildPackage() throws Exception {
    String[] sources = {
        "child_pkg.proto"
    };
    invokeCompiler(sources, "--named_files_only");

    String[] outputs = {
        "com/squareup/wire/protos/ChildPackage.java",
    };
    assertOutputs(outputs);
  }

  @Test public void testAllTypes() throws Exception {
    String[] sources = {
        "all_types.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/alltypes/AllTypes.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testEdgeCases() throws Exception {
    String[] sources = {
        "edge_cases.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/edgecases/NoFields.java",
        "com/squareup/wire/protos/edgecases/OneField.java",
        "com/squareup/wire/protos/edgecases/OneBytesField.java",
        "com/squareup/wire/protos/edgecases/Recursive.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testUnknownFields() throws Exception {
    String[] sources = {
        "unknown_fields.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/unknownfields/VersionOne.java",
        "com/squareup/wire/protos/unknownfields/VersionTwo.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testCustomOptions() throws Exception {
    String[] sources = {
        "custom_options.proto",
        "option_redacted.proto"
    };
    invokeCompiler(sources, "--named_files_only");

    String[] outputs = {
        "com/squareup/wire/protos/custom_options/FooBar.java",
        "com/squareup/wire/protos/custom_options/MessageWithOptions.java",
    };
    assertOutputs(outputs);
  }

  @Test public void testCustomOptionsNoOptions() throws Exception {
    String[] sources = {
        "custom_options.proto",
        "option_redacted.proto"
    };
    invokeCompiler(sources, "--excludes=google.protobuf.*", "--named_files_only");

    String[] outputs = {
        "com/squareup/wire/protos/custom_options/FooBar.java",
        "com/squareup/wire/protos/custom_options/MessageWithOptions.java"
    };
    assertOutputs(outputs, ".noOptions");
  }

  @Test public void testRedacted() throws Exception {
    String[] sources = {
        "redacted_test.proto",
        "option_redacted.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/redacted/NotRedacted.java",
        "com/squareup/wire/protos/redacted/Redacted.java",
        "com/squareup/wire/protos/redacted/RedactedChild.java",
        "com/squareup/wire/protos/redacted/RedactedCycleA.java",
        "com/squareup/wire/protos/redacted/RedactedCycleB.java",
        "com/squareup/wire/protos/redacted/RedactedExtension.java",
        "com/squareup/wire/protos/redacted/RedactedRepeated.java",
        "com/squareup/wire/protos/redacted/RedactedRequired.java",
    };
    assertOutputs(outputs);
  }

  @Test public void testNoRoots() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources);

    String[] outputs = {
        "com/squareup/wire/protos/roots/A.java",
        "com/squareup/wire/protos/roots/B.java",
        "com/squareup/wire/protos/roots/C.java",
        "com/squareup/wire/protos/roots/D.java",
        "com/squareup/wire/protos/roots/E.java",
        "com/squareup/wire/protos/roots/G.java",
        "com/squareup/wire/protos/roots/H.java",
        "com/squareup/wire/protos/roots/I.java",
        "com/squareup/wire/protos/roots/J.java",
        "com/squareup/wire/protos/roots/K.java",
    };
    assertOutputs(outputs);
  }

  @Test public void testExcludes() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources,
        "--includes=squareup.protos.roots.A",
        "--excludes=squareup.protos.roots.B");

    String[] outputs = {
        "com/squareup/wire/protos/roots/A.java",
        "com/squareup/wire/protos/roots/D.java"
    };
    assertOutputs(outputs, ".pruned");
  }

  @Test public void testRootsA() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources, "--includes=squareup.protos.roots.A");

    String[] outputs = {
        "com/squareup/wire/protos/roots/A.java",
        "com/squareup/wire/protos/roots/B.java",
        "com/squareup/wire/protos/roots/C.java",
        "com/squareup/wire/protos/roots/D.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testRootsB() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources, "--includes=squareup.protos.roots.B");

    String[] outputs = {
        "com/squareup/wire/protos/roots/B.java",
        "com/squareup/wire/protos/roots/C.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testRootsE() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources, "--includes=squareup.protos.roots.E");

    String[] outputs = {
        "com/squareup/wire/protos/roots/E.java",
        "com/squareup/wire/protos/roots/G.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testRootsH() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources, "--includes=squareup.protos.roots.H");

    String[] outputs = {
        "com/squareup/wire/protos/roots/E.java",
        "com/squareup/wire/protos/roots/H.java"
    };
    assertOutputs(outputs, ".pruned");
  }

  @Test public void testRootsI() throws Exception {
    String[] sources = {
        "roots.proto"
    };
    invokeCompiler(sources, "--includes=squareup.protos.roots.I");

    String[] outputs = {
        "com/squareup/wire/protos/roots/I.java",
        "com/squareup/wire/protos/roots/J.java",
        "com/squareup/wire/protos/roots/K.java"
    };
    assertOutputs(outputs);
  }

  @Test public void testDryRun() throws Exception {
    String[] sources = {
        "service_root.proto"
    };
    invokeCompiler(sources, "--includes=squareup.wire.protos.roots.TheService", "--dry_run", "--quiet");

    assertThat(logger.getLog()).isEqualTo(""
        + testDir.getAbsolutePath() + " com.squareup.wire.protos.roots.TheRequest\n"
        + testDir.getAbsolutePath() + " com.squareup.wire.protos.roots.TheResponse\n");
  }

  @Test public void noFiles() throws Exception {
    String[] sources = new String[0];
    invokeCompiler(sources);

    assertThat(getPaths()).isNotEmpty();
  }

  private void invokeCompiler(String[] sources, String... extraArgs) throws Exception {
    List<String> args = new ArrayList<>();
    args.add("--proto_path=../wire-runtime/src/test/proto");
    args.add("--java_out=" + testDir.getAbsolutePath());
    Collections.addAll(args, extraArgs);
    Collections.addAll(args, sources);

    logger = new StringWireLogger();
    FileSystem fs = FileSystems.getDefault();
    WireCompiler compiler = WireCompiler.forArgs(fs, logger, args.toArray(new String[args.size()]));
    compiler.compile();
  }

  private void assertOutputs(String[] outputs) throws IOException {
    assertOutputs(outputs, "");
  }

  private void assertOutputs(String[] outputs, String suffix) throws IOException {
    List<String> filesAfter = getPaths();
    assertThat(filesAfter.size())
        .overridingErrorMessage(filesAfter.toString())
        .isEqualTo(outputs.length);

    for (String output : outputs) {
      assertFilesMatch(testDir, output, suffix);
    }
  }

  /** Returns all paths within {@code root}, and relative to {@code root}. */
  private List<String> getPaths() {
    List<String> paths = new ArrayList<>();
    getPathsRecursive(testDir.getAbsoluteFile(), "", paths);
    return paths;
  }

  private void getPathsRecursive(File base, String path, List<String> paths) {
    File file = new File(base, path);

    String[] children = file.list();
    if (children == null) return;

    for (String child : children) {
      File childFile = new File(file, child);
      if (childFile.isFile()) {
        paths.add(path + child);
      } else {
        getPathsRecursive(base, path + child + "/", paths);
      }
    }
  }

  private void assertFilesMatch(File outputDir, String path, String suffix) throws IOException {
    // Compare against file with suffix if present
    File expectedFile = new File("../wire-runtime/src/test/proto-java/" + path + suffix);
    if (expectedFile.exists()) {
      System.out.println("Comparing against expected output " + expectedFile.getName());
    } else {
      expectedFile = new File("../wire-runtime/src/test/proto-java/" + path);
    }
    File actualFile = new File(outputDir, path);
    assertFilesMatch(expectedFile, actualFile);
  }

  private void assertFilesMatch(File expectedFile, File actualFile) throws IOException {
    String expected;
    try (Source source = Okio.source(expectedFile)) {
      expected = Okio.buffer(source).readUtf8();
    }

    String actual;
    try (Source source = Okio.source(actualFile)) {
      actual = Okio.buffer(source).readUtf8();
    }

    // Normalize CRLF -> LF
    expected = expected.replace("\r\n", "\n");
    actual = actual.replace("\r\n", "\n");
    assertThat(actual).isEqualTo(expected);
  }
}

File: wire-compiler/src/test/java/com/squareup/wire/WireCompilerTest.kt
/*
 * Copyright (C) 2013 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("UsePropertyAccessSyntax")

package com.squareup.wire

import assertk.all
import assertk.assertThat
import assertk.assertions.contains
import assertk.assertions.containsExactly
import assertk.assertions.containsOnly
import assertk.assertions.isEmpty
import assertk.assertions.isEqualTo
import assertk.assertions.isFalse
import assertk.assertions.isNotNull
import assertk.assertions.isNull
import assertk.assertions.isTrue
import assertk.assertions.prop
import assertk.assertions.size
import com.squareup.wire.kotlin.EnumMode
import com.squareup.wire.kotlin.RpcCallStyle
import com.squareup.wire.kotlin.RpcRole
import com.squareup.wire.schema.CustomTarget
import com.squareup.wire.schema.EmptyWireLogger
import com.squareup.wire.schema.JavaTarget
import com.squareup.wire.schema.KotlinTarget
import com.squareup.wire.schema.Location
import com.squareup.wire.schema.SwiftTarget
import com.squareup.wire.schema.WireRun
import okio.Path.Companion.toPath
import okio.buffer
import okio.fakefilesystem.FakeFileSystem
import org.junit.Test

class WireCompilerTest {
  private val fileSystem = FakeFileSystem()

  @Test
  fun compileExecutes() {
    val outputDirectory = "/out".toPath()

    fileSystem.appendingSink("baz.proto".toPath()).buffer().use {
      it.writeUtf8(
        """
          |syntax = "proto2";
          |
          |message Player {
          |  optional string name = 1;
          |}
          |
        """.trimMargin(),
      )
    }
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--proto_path=/",
      "--java_out=$outputDirectory",
    )
    wireCompiler.compile()

    assertThat(fileSystem.listRecursively("/out".toPath())).containsExactly(outputDirectory / "Player.java")
  }

  @Test
  fun default() {
    val wireCompiler = WireCompiler.forArgs(fileSystem, EmptyWireLogger(), "--java_out=out/")
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::sourcePath).isEmpty()
      prop(WireRun::protoPath).isEmpty()
      prop(WireRun::treeShakingRoots).containsExactly("*")
      prop(WireRun::treeShakingRubbish).isEmpty()
      prop(WireRun::moves).isEmpty()
      prop(WireRun::sinceVersion).isNull()
      prop(WireRun::untilVersion).isNull()
      prop(WireRun::onlyVersion).isNull()
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "out/",
        ),
      )
      prop(WireRun::modules).isEmpty()
      prop(WireRun::permitPackageCycles).isFalse()
      prop(WireRun::loadExhaustively).isFalse()
      prop(WireRun::escapeKotlinKeywords).isFalse()
      prop(WireRun::eventListeners).isEmpty()
      prop(WireRun::rejectUnusedRootsOrPrunes).isTrue()
      prop(WireRun::opaqueTypes).isEmpty()
    }
  }

  @Test // TODO(Benoit) Delete? Nobody wanna maintain such a test
  fun allFlags() {
    fileSystem.appendingSink("foo.proto".toPath())
    fileSystem.appendingSink("bar.proto".toPath())
    fileSystem.appendingSink("proto.include".toPath())
      .buffer().use {
        it.writeUtf8("foo.proto\n")
        it.writeUtf8("bar.proto\n")
      }
    fileSystem.appendingSink("proto_manifest.yaml".toPath())
      .buffer().use {
        it.writeUtf8(
          """
          |a: {}
          |b:
          |  dependencies:
          |   - a
          |
          """.trimMargin(),
        )
      }

    val wireCompiler = WireCompiler.forArgs(
      fileSystem, EmptyWireLogger(),
      "--proto_path=/",
      "--java_out=java_out",
      "--kotlin_out=kotlin_out",
      "--swift_out=swift_out",
      // "--custom_out=custom_out",
      // "--schema_handler_factory_class=schema_handler_factory_class",
      "--files=proto.include",
      "--includes=includes",
      "--excludes=excludes",
      "--experimental-module-manifest=proto_manifest.yaml",
      // "--event_listener_factory_class=event_listener_factory_class",
      // "--logger_factory_class=logger_factory_class",
      "--android",
      "--android-annotations",
      "--compact",
      "--skip_declared_options",
      "--skip_applied_options",
      "--permit_package_cycles",
      "--load_exhaustively",
      "--java_interop",
      "--dry_run",
      "--kotlin_box_oneofs_min_size=67",
      "--no_java_exclusive",
      "--no_kotlin_exclusive",
      "--no_swift_exclusive",
      "--kotlin_rpc_call_style=blocking",
      "--kotlin_rpc_role=server",
      "--kotlin_single_method_services",
      "--kotlin_name_suffix=kotlin_name_suffix",
      "--kotlin_builders_only",
      "--kotlin_escape_keywords",
      "--emit_proto_reader_32",
      "--kotlin_enum_mode=sealed_class",
      // "--custom_option=a,1",
      // "--custom_option=b,2",
      "--opaque_types=opaque_types",
      "--ignore_unused_roots_and_prunes",
      "--kotlin_explicit_streaming_calls",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::sourcePath).containsExactly(Location.get("foo.proto"), Location.get("bar.proto"))
      prop(WireRun::protoPath).containsExactly(Location.get("/"))
      prop(WireRun::treeShakingRoots).containsExactly("includes")
      prop(WireRun::treeShakingRubbish).containsExactly("excludes")
      // There's no flag to define type moves.
      prop(WireRun::moves).isEmpty()
      prop(WireRun::sinceVersion).isNull()
      prop(WireRun::untilVersion).isNull()
      prop(WireRun::onlyVersion).isNull()
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          includes = listOf("*"),
          excludes = listOf(),
          exclusive = false,
          outDirectory = "java_out",
          android = true,
          androidAnnotations = true,
          compact = true,
          emitDeclaredOptions = false,
          emitAppliedOptions = false,
          buildersOnly = false,
        ),
        KotlinTarget(
          includes = listOf("*"),
          excludes = listOf(),
          exclusive = false,
          outDirectory = "kotlin_out",
          android = true,
          javaInterop = true,
          emitDeclaredOptions = false,
          emitAppliedOptions = false,
          rpcCallStyle = RpcCallStyle.BLOCKING,
          rpcRole = RpcRole.SERVER,
          singleMethodServices = true,
          boxOneOfsMinSize = 67,
          nameSuffix = "kotlin_name_suffix",
          buildersOnly = true,
          escapeKotlinKeywords = true,
          enumMode = EnumMode.SEALED_CLASS,
          emitProtoReader32 = true,
          mutableTypes = false,
          explicitStreamingCalls = true,
        ),
        SwiftTarget(
          includes = listOf("*"),
          excludes = listOf(),
          exclusive = false,
          outDirectory = "swift_out",
        ),
      )
      prop(WireRun::modules).isEqualTo(
        mapOf("a" to WireRun.Module(), "b" to WireRun.Module(dependencies = setOf("a"))),
      )
      prop(WireRun::permitPackageCycles).isTrue()
      prop(WireRun::loadExhaustively).isTrue()
      prop(WireRun::escapeKotlinKeywords).isFalse()
      prop(WireRun::eventListeners).isEmpty()
      prop(WireRun::rejectUnusedRootsOrPrunes).isFalse()
      prop(WireRun::opaqueTypes).containsExactly("opaque_types")
    }
  }

  @Test
  fun protoLocations() {
    fileSystem.appendingSink("foo.proto".toPath())
    fileSystem.appendingSink("bar.proto".toPath())
    fileSystem.appendingSink("proto.include".toPath())
      .buffer().use {
        it.writeUtf8("foo.proto\n")
        it.writeUtf8("bar.proto\n")
      }
    fileSystem.appendingSink("baz.proto".toPath())

    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--proto_path=/",
      "--java_out=java_out",
      "--files=proto.include",
      "baz.proto",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::sourcePath).containsExactly(Location.get("foo.proto"), Location.get("bar.proto"), Location.get("baz.proto"))
      prop(WireRun::protoPath).containsExactly(Location.get("/"))
      prop(WireRun::treeShakingRoots).containsExactly("*")
      prop(WireRun::treeShakingRubbish).isEmpty()
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "java_out",
        ),
      )
    }
  }

  @Test
  fun treeShaking() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--kotlin_out=kotlin_out",
      "--includes=includes1,includes2,includes3",
      "--excludes=excludes1,excludes2",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::treeShakingRoots).containsExactly("includes1", "includes2", "includes3")
      prop(WireRun::treeShakingRubbish).containsExactly("excludes1", "excludes2")
      prop(WireRun::targets).containsExactly(
        KotlinTarget(
          outDirectory = "kotlin_out",
        ),
      )
    }
  }

  @Test
  fun manifest() {
    fileSystem.appendingSink("proto_manifest.yaml".toPath())
      .buffer().use {
        it.writeUtf8(
          """
          |a: {}
          |b:
          |  dependencies:
          |   - a
          |
          """.trimMargin(),
        )
      }

    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--java_out=java_out",
      "--experimental-module-manifest=proto_manifest.yaml",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "java_out",
        ),
      )
      prop(WireRun::modules).isEqualTo(
        mapOf("a" to WireRun.Module(), "b" to WireRun.Module(dependencies = setOf("a"))),
      )
    }
  }

  @Test
  fun loadExhaustively() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--java_out=java_out",
      "--load_exhaustively",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "java_out",
        ),
      )
      prop(WireRun::loadExhaustively).isTrue()
    }
  }

  @Test
  fun permitPackageCycles() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--java_out=java_out",
      "--permit_package_cycles",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "java_out",
        ),
      )
      prop(WireRun::permitPackageCycles).isTrue()
    }
  }

  @Test
  fun dryRun() {
    val outputDirectory = "/out".toPath()

    fileSystem.appendingSink("baz.proto".toPath()).buffer().use {
      it.writeUtf8(
        """
          |syntax = "proto2";
          |
          |message Player {
          |  optional string name = 1;
          |}
          |
        """.trimMargin(),
      )
    }
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--proto_path=/",
      "--java_out=$outputDirectory",
      "--dry_run",
    )
    wireCompiler.compile()

    // Dry run didn't write on disk.
    assertThat(fileSystem.listRecursively("/out".toPath())).isEmpty()
  }

  @Test
  fun opaqueTypes() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--java_out=java_out",
      "--opaque_types=opaque_types",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "java_out",
        ),
      )
      prop(WireRun::opaqueTypes).containsExactly("opaque_types")
    }
  }

  @Test
  fun ignoreUnusedRootsAndPrunes() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--java_out=java_out",
      "--ignore_unused_roots_and_prunes",
    )
    val wireRun = wireCompiler.createRun()
    assertThat(wireRun).all {
      prop(WireRun::targets).containsExactly(
        JavaTarget(
          outDirectory = "java_out",
        ),
      )
      prop(WireRun::rejectUnusedRootsOrPrunes).isFalse()
    }
  }

  @Test
  fun allTargetsAndAllOptions() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem, EmptyWireLogger(),
      "--proto_path=/",
      "--java_out=java_out",
      "--kotlin_out=kotlin_out",
      "--swift_out=swift_out",
      "--custom_out=custom_out",
      "--schema_handler_factory_class=schema_handler_factory_class",
      "--android",
      "--android-annotations",
      "--compact",
      "--skip_declared_options",
      "--skip_applied_options",
      "--permit_package_cycles",
      "--java_interop",
      "--kotlin_box_oneofs_min_size=67",
      "--no_java_exclusive",
      "--no_kotlin_exclusive",
      "--no_swift_exclusive",
      "--kotlin_rpc_call_style=blocking",
      "--kotlin_rpc_role=server",
      "--kotlin_single_method_services",
      "--kotlin_name_suffix=kotlin_name_suffix",
      "--kotlin_builders_only",
      "--kotlin_escape_keywords",
      "--emit_proto_reader_32",
      "--kotlin_enum_mode=sealed_class",
      "--custom_option=a,1",
      "--custom_option=b,2",
      "--kotlin_explicit_streaming_calls",
    )
    val wireRun = wireCompiler.createRun()
    val targets = wireRun.targets
    assertThat(targets.filterIsInstance<JavaTarget>().single()).isEqualTo(
      JavaTarget(
        includes = listOf("*"),
        excludes = listOf(),
        exclusive = false,
        outDirectory = "java_out",
        android = true,
        androidAnnotations = true,
        compact = true,
        emitDeclaredOptions = false,
        emitAppliedOptions = false,
        buildersOnly = false,
      ),
    )
    assertThat(targets.filterIsInstance<KotlinTarget>().single()).isEqualTo(
      KotlinTarget(
        includes = listOf("*"),
        excludes = listOf(),
        exclusive = false,
        outDirectory = "kotlin_out",
        android = true,
        javaInterop = true,
        emitDeclaredOptions = false,
        emitAppliedOptions = false,
        rpcCallStyle = RpcCallStyle.BLOCKING,
        rpcRole = RpcRole.SERVER,
        singleMethodServices = true,
        boxOneOfsMinSize = 67,
        nameSuffix = "kotlin_name_suffix",
        buildersOnly = true,
        escapeKotlinKeywords = true,
        enumMode = EnumMode.SEALED_CLASS,
        emitProtoReader32 = true,
        mutableTypes = false,
        explicitStreamingCalls = true,
      ),
    )
    assertThat(targets.filterIsInstance<SwiftTarget>().single()).isEqualTo(
      SwiftTarget(
        includes = listOf("*"),
        excludes = listOf(),
        exclusive = false,
        outDirectory = "swift_out",
      ),
    )
    val customTarget = targets.filterIsInstance<CustomTarget>().single()
    assertThat(customTarget).all {
      prop(CustomTarget::includes).containsExactly("*")
      prop(CustomTarget::excludes).isEmpty()
      prop(CustomTarget::exclusive).isTrue()
      prop(CustomTarget::outDirectory).isEqualTo("custom_out")
      prop(CustomTarget::options).containsOnly("a" to "1", "b" to "2")
      prop(CustomTarget::schemaHandlerFactory).isNotNull()
    }
    assertThat(targets.size).isEqualTo(4)
  }

  @Test
  fun customOutput() {
    val wireCompiler = WireCompiler.forArgs(
      fileSystem,
      EmptyWireLogger(),
      "--proto_path=/",
      "--custom_out=custom_out",
      "--schema_handler_factory_class=schema_handler_factory_class",
      "--custom_option=a,1",
      "--custom_option=b,2",
    )
    val wireRun = wireCompiler.createRun()
    val target = wireRun.targets.single() as CustomTarget
    assertThat(target.includes).containsExactly("*")
    assertThat(target.excludes).isEmpty()
    assertThat(target.exclusive).isTrue()
    assertThat(target.options).isEqualTo(mapOf("a" to "1", "b" to "2"))
    assertThat(target.schemaHandlerFactory).isNotNull()
  }
}
